### ISSUE TITLE
Wrong relative path passed to lint.Run()

### ISSUE THREAD

**Issue by magnusbaeck** [carsongee/pytest-pylint#26]
After #23, it seems pytest-pylint no longer passes the absolute paths to the files to lint but instead relative paths. In our case the paths are relative to the wrong directory, leading to this kind of error:

```
__________________________ [pylint] axis/__init__.py ___________________________
F:  1, 0: No module py in axis/__init__ (fatal)
```

This happens when I run `py.test src/axis --pylint`. On closer inspection pytest-pylint apparently passes `axis/__init__.py` to lint.Run(), but this assumes that the current directory is src, which it clearly isn't. If I step into the src directory and run `py.test axis --pylint` things work fine. This seems like a regression.

**Comment by carsongee** [comment]
Thanks for reporting the issue!  I'm happy to try and figure this out, but I can't reproduce the issue so far with my little setup, so I'm not sure what is going on.  I'm guessing it is something to do with mine not being quite complex enough, here is what I have:
```
(tmp-93b2d225d5f78cc1)#[13:07:30] carson@carson-lt00:~/tmp/pytest_issue $
find .
.
./src
./src/axis
./src/axis/__init__.py
(tmp-93b2d225d5f78cc1)#[13:07:32] carson@carson-lt00:~/tmp/pytest_issue $
cat src/axis/__init__.py
#sadflkasdf

def blah(args=[]):
    print(args)


# asdflaksdflkjasdfljkasdflkjasdkljfasldkjfakljsdfklasdflkasdlkjfaslkjdflkasdflk;asfdkl;
(tmp-93b2d225d5f78cc1)#[13:07:35] carson@carson-lt00:~/tmp/pytest_issue $
py.test src/axis --pylint
============================================================ test session starts ============================================================
platform darwin -- Python 2.7.10, pytest-3.0.6, py-1.4.32, pluggy-0.4.0
rootdir: /Users/carson/tmp/pytest_issue, inifile:
plugins: pylint-0.7.0
collected 1 items
-----------------------------------------------------------------
Linting files
No config file found, using default configuration
-----------------------------------------------------------------

src/axis/__init__.py F

================================================================= FAILURES ==================================================================
_______________________________________________________ [pylint] src/axis/__init__.py _______________________________________________________
C:  4, 0: Unnecessary parens after 'print' keyword (superfluous-parens)
C:  1, 0: Missing module docstring (missing-docstring)
W:  3, 0: Dangerous default value [] as argument (dangerous-default-value)
C:  3, 0: Missing function docstring (missing-docstring)
========================================================= 1 failed in 0.05 seconds ==========================================================
(tmp-93b2d225d5f78cc1)#[13:07:38] carson@carson-lt00:~/tmp/pytest_issue $
```

which all seems to be the correct behavior, I also note that you have `__________________________ [pylint] axis/__init__.py ___________________________` in your output, but I have `_______________________________________________________ [pylint] src/axis/__init__.py _______________________________________________________` so I'm not sure what is going on, any help you can provide is appreciated.

**Comment by magnusbaeck** [comment]
I'll look into a minimal recipe for reproducing the problem tomorrow. I run Python 3.4/3.5 on Linux but that shouldn't make a difference here, I think.

**Comment by eddie-dunn** [comment]
First of all, thank you for the swift response. I will try to supply some more info to help you get started.

You should be able to reproduce the problem with something similar to this layout:
```
% tree  
.
└── src
    ├── axis
    │   ├── app
    │   │   └── __init__.py
    │   └── __init__.py
    └── setup.py

3 directories, 3 files

```
All files, including `setup.py`, are empty.

Running:
```
% py.test --pylint src/
=========================== test session starts ===========================
platform linux -- Python 3.6.0, pytest-3.0.6, py-1.4.32, pluggy-0.4.0
rootdir: /tmp/testfoo/src, inifile: 
plugins: pylint-0.7.0
collected 3 items 
-----------------------------------------------------------------
Linting files
No config file found, using default configuration
-----------------------------------------------------------------

src/setup.py F
src/axis/__init__.py F
src/axis/app/__init__.py F

================================ FAILURES =================================
____________________________ [pylint] setup.py ____________________________
F:  1, 0: No module named setup.py (fatal)
________________________ [pylint] axis/__init__.py ________________________
F:  1, 0: No module named axis/__init__.py (fatal)
______________________ [pylint] axis/app/__init__.py ______________________
F:  1, 0: No module named axis/app/__init__.py (fatal)
======================== 3 failed in 0.07 seconds =========================
```

Now here's the fun part.

```
% mv src/setup.py .
% py.test --pylint src/
=========================== test session starts ===========================
platform linux -- Python 3.6.0, pytest-3.0.6, py-1.4.32, pluggy-0.4.0
rootdir: /tmp/testfoo, inifile: 
plugins: pylint-0.7.0
collected 2 items 
-----------------------------------------------------------------
Linting files
No config file found, using default configuration
-----------------------------------------------------------------

src/axis/__init__.py .
src/axis/app/__init__.py .

======================== 2 passed in 0.07 seconds =========================
```
Tests pass!
```
% mv setup.py ../
% py.test --pylint src/
=========================== test session starts ===========================
platform linux -- Python 3.6.0, pytest-3.0.6, py-1.4.32, pluggy-0.4.0
rootdir: /tmp, inifile: 
plugins: pylint-0.7.0
collected 2 items 
-----------------------------------------------------------------
Linting files
No config file found, using default configuration
-----------------------------------------------------------------

src/axis/__init__.py F
src/axis/app/__init__.py F

================================ FAILURES =================================
__________________ [pylint] testfoo/src/axis/__init__.py __________________
F:  1, 0: No module named testfoo/src/axis/__init__.py (fatal)
________________ [pylint] testfoo/src/axis/app/__init__.py ________________
F:  1, 0: No module named testfoo/src/axis/app/__init__.py (fatal)
======================== 2 failed in 0.07 seconds =========================
```
Tests fail again!
```
% mv ../setup.py ~/
% py.test --pylint src/
=========================== test session starts ===========================
platform linux -- Python 3.6.0, pytest-3.0.6, py-1.4.32, pluggy-0.4.0
rootdir: /tmp/testfoo, inifile: 
plugins: pylint-0.7.0
collected 2 items 
-----------------------------------------------------------------
Linting files
No config file found, using default configuration
-----------------------------------------------------------------

src/axis/__init__.py .
src/axis/app/__init__.py .

======================== 2 passed in 0.07 seconds =========================
```
Tests pass again!

In case the working directory is important:
```
% pwd
/tmp/testfoo
```

### Summary
If I move  `setup.py` from the `src` folder to `.`, it works. If I move `setup.py` to `../` I get the same `No module named ... (fatal)` error again. If I move `setup.py` somewhere else entirely it magically works again.

**Comment by carsongee** [comment]
I've done some digging here, but don't have the best answers.  Apparently the `rootdir` can move with where it finds the `setup.py` file from pytest which is what `pytest-pylint` is keying off to pass the right relative paths into PyLint.  This is what seems to be causing the issues.  The problem is, I don't really know what the right behavior is for this project.  I'd rather not key off the current working directory, because that can cause other issues, but obviously this isn't the best either.  From my testing, it appears that if you just leave the path off entirely it finds the right rootdir as well:
```
(tmp-71d2061de966ed72)#[17:13:42] carson@carson-lt00:~/tmp/testo $
find .
.
./src
./src/axis
./src/axis/__init__.py
./src/axis/app
./src/axis/app/__init__.py
./src/setup.py
(tmp-71d2061de966ed72)#[17:13:44] carson@carson-lt00:~/tmp/testo $
pytest --pylint src/ -s
=========================================================================================== test session starts ============================================================================================
platform darwin -- Python 2.7.10, pytest-3.0.6, py-1.4.32, pluggy-0.4.0
rootdir: /Users/carson/tmp/testo/src, inifile:
plugins: pylint-0.7.0
collected 3 items
-----------------------------------------------------------------
Linting files
No config file found, using default configuration
-----------------------------------------------------------------

src/setup.py F
src/axis/__init__.py F
src/axis/app/__init__.py F

================================================================================================= FAILURES =================================================================================================
____________________________________________________________________________________________ [pylint] setup.py _____________________________________________________________________________________________
F:  1, 0: No module py in setup (fatal)
________________________________________________________________________________________ [pylint] axis/__init__.py _________________________________________________________________________________________
F:  1, 0: No module named axis/__init__.py (fatal)
______________________________________________________________________________________ [pylint] axis/app/__init__.py _______________________________________________________________________________________
F:  1, 0: No module named axis/app/__init__.py (fatal)
========================================================================================= 3 failed in 0.06 seconds =========================================================================================
(tmp-71d2061de966ed72)#[17:13:53] carson@carson-lt00:~/tmp/testo $
pytest --pylint  -s
=========================================================================================== test session starts ============================================================================================
platform darwin -- Python 2.7.10, pytest-3.0.6, py-1.4.32, pluggy-0.4.0
rootdir: /Users/carson/tmp/testo, inifile:
plugins: pylint-0.7.0
collected 3 items
-----------------------------------------------------------------
Linting files
No config file found, using default configuration
-----------------------------------------------------------------

src/setup.py .
src/axis/__init__.py .
src/axis/app/__init__.py F

================================================================================================= FAILURES =================================================================================================
____________________________________________________________________________________ [pylint] src/axis/app/__init__.py _____________________________________________________________________________________
C:  2, 0: Unnecessary parens after 'print' keyword (superfluous-parens)
C:  1, 0: Missing module docstring (missing-docstring)
C:  1, 0: Invalid function name "hi" (invalid-name)
C:  1, 0: Missing function docstring (missing-docstring)
==================================================================================== 1 failed, 2 passed in 0.07 seconds ====================================================================================
```

When I get a little more time, I'll ask the pytest people if they know what is going on here, but for now, the best I can offer is the workaround of not passing the path to `src` or executing the pylint test from the `src` directory.

**Comment by eddie-dunn** [comment]
Thanks, we worked around the issue by using the method you suggested (i.e., making `src` the current working dir).

Feel free to ping me if you need help testing/solving the problem.

**Comment by quartox** [comment]
Bumping. I am getting the same problem and the workarounds above aren't working for me. Specifically, my user home directory is always the `rootdir` regardless of existence of `setup.py` and what the current working directory is when I run `pytest --pylint`.

My current workaround is to just disable `fatal` and assume my unit tests would catch any truly fatal errors.

**Comment by Meltz014** [comment]
Bumping again.  I see this issue and i think it's due to performing a chdir in my `pytest_configure()` hook in `conftest.py`

**Comment by carsongee** [comment]
This was from a long ago and there have been many new pytests. I'm hoping that has meant this is no longer relevant since it is more of a pytest problem. That said, feel free to re-open if needed
